leopard/leopard.nim

```diff
@@ -0,0 +1,263 @@
+## Nim-Leopard
+## Copyright (c) 2022 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import pkg/upraises
+push: {.upraises: [].}
+
+{.deadCodeElim: on.}
+
+import pkg/stew/results
+
+import ./wrapper
+import ./utils
+
+export wrapper, results
+
+const
+  BuffMultiples* = 64
+
+type
+  LeoBufferPtr* = ptr UncheckedArray[byte]
+
+  LeoCoderKind* {.pure.} = enum
+    Encoder,
+    Decoder
+
+  Leo* = object of RootObj
+    bufSize*: int                         # size of the buffer in multiples of 64
+    buffers*: int                         # total number of data buffers (K)
+    parity*: int                          # total number of parity buffers (M)
+    dataBufferPtr: seq[LeoBufferPtr]      # buffer where data is copied before encoding
+    workBufferCount: int                  # number of parity work buffers
+    workBufferPtr: seq[LeoBufferPtr]      # buffer where parity is copied before encoding
+    case kind: LeoCoderKind
+    of LeoCoderKind.Decoder:
+      decodeBufferCount: int              # number of decoding work buffers
+      decodeBufferPtr: seq[LeoBufferPtr]  # work buffer used for decoding
+    of LeoCoderKind.Encoder:
+      discard
+
+  LeoEncoder* = object of Leo
+  LeoDecoder* = object of Leo
+
+func encode*(
+  self: var LeoEncoder,
+  data,
+  parity: var openArray[seq[byte]]): Result[void, cstring] =
+  ## Encode a list of buffers in `data` into a number of `bufSize` sized
+  ## `parity` buffers
+  ##
+  ## `data`   - list of original data `buffers` of size `bufSize`
+  ## `parity` - list of parity `buffers` of size `bufSize`
+  ##
+
+  if data.len != self.buffers:
+    return err("Number of data buffers should match!")
+
+  if parity.len != self.parity:
+    return err("Number of parity buffers should match!")
+
+  # zero encode work buffer to avoid corrupting with previous run
+  for i in 0..<self.workBufferCount:
+    zeroMem(self.workBufferPtr[i], self.bufSize)
+
+  # copy data into aligned buffer
+  for i in 0..<data.len:
+    copyMem(self.dataBufferPtr[i], addr data[i][0], self.bufSize)
```

Comment:
Sorry if the question seems dumb, it's more about my wanting/needing to understand better what happens "under the hood" with the memory when data is added to or removed from a `seq` vs. when a seq is initialized with all zeroed entries.

Reply:
When dealing with "array types" (array, seq), it's safe to assume that memory is laid out sequentially.

In `C/C++` there is direct relationship of pointers and arrays, an array is nothing more than a pointer to a contiguous memory segment and in most cases the two can be used interchangeably. For example, the encode/decode `C` Api functions could have been declared as `uint8_t data[][]`, instead of `void** data`.

The assumption we're making here is that each `seq[byte]` in the `seq[seq[byte]]` points to a contiguous sequence of memory, for which we get a pointer and then copy the bytes to another segment of memory pointed to by `dataBufferPtr[i]`. This is safe because of both how Nim promises to keep this consistent (at least for the C/C++ backends) and how `C/C++` arrays work. (This seems to be a good high level overview of pointers and arrays https://www.programiz.com/c-programming/c-pointers-arrays)

To give a few more examples, let's look at our `data: seq[seq[byte]]` again.

- `addr data[0]` would get the address of the `seq` located in position `0` (a seq is really just a struct containing a length and a `C` array)
- `addr data[0][0]` would get the address of the first byte stored in the array of the `seq` found at position 0 of `data`

Now, the difference between an `array of arrays` (or `seq[seq[byte]]`) and an array of pointers (or `seq[ptr pointer]`) is that the latter can point to **any** location on the heap (or even the stack), where that segments of memory exist, which is exactly what we do when we're allocating with `leoAlloc` (and subsequently `malloc`, etc...) and storing the pointer to it in `dataBufferPtr` (and friends).